### PR TITLE
[inductor] fix sdpa pattern matcher issue when scale is not default 

### LINKED ISSUE #151098: precision error for attention like-operation
### 🐛 Describe the bug

tlparse file: [dedicated_log_torch_trace_ve9nm2rv.log](https://github.com/user-attachments/files/19702029/dedicated_log_torch_trace_ve9nm2rv.log)
The result before and after torch.compile have a big difference:
```
Maximum error between out_normal and out_opt: 2.276662826538086 # first try
Maximum error between out_normal and out_opt: 2.2177655696868896 # second try
Maximum error between out_normal and out_opt: 2.2525100708007812 # third try
```
Here is the source. maybe caused by indcuctor/sfdp optimization:
```python
import torch

class Model(torch.nn.Module):

    def __init__(self):
        super().__init__()
        self.inv_scale = 1.0 / 8 ** 0.5
        self.query = torch.nn.Linear(64, 64)
        self.key = torch.nn.Linear(64, 64)
        self.value = torch.nn.Linear(64, 64)

    def forward(self, x1, attn_mask):
        q = self.query(x1).permute([0, 2, 1, 3])
        k = self.key(x1).permute([0, 2, 1, 3])
        v = self.value(x1).permute([0, 2, 1, 3])
        t1 = torch.matmul(q, k.transpose(-2, -1))
        t2 = t1.div(self.inv_scale)
        t3 = t2 + attn_mask
        t4 = t3.softmax(dim=-1)
        t5 = t4.matmul(v)
        return t5


func = Model().to('cpu')


x1 = torch.randn(1, 16, 64, 64)

attn_mask = torch.zeros(1, 1, 16, 16)

test_inputs = [x1, attn_mask]

opt = torch.compile(func)

out_normal = func(x1, attn_mask)
out_opt = opt(x1, attn_mask)

# Calculate maximum error between out_normal and out_opt
max_error = torch.max(torch.abs(out_normal - out_opt))
print(f"Maximum error between out_normal and out_opt: {max_error.item()}")
```

### Error logs

_No response_

### Versions

```
Collecting environment information...
PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Feb  4 2025, 14:57:36) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               18
On-line CPU(s) list:                  0-17
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Core(TM) Ultra 5 125H
CPU family:                           6
Model:                                170
Thread(s) per core:                   2
Core(s) per socket:                   9
Socket(s):                            1
Stepping:                             4
BogoMIPS:                             5990.39
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology tsc_reliable nonstop_tsc cpuid pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves avx_vnni umip waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize flush_l1d arch_capabilities
Virtualization:                       VT-x
Hypervisor vendor:                    Microsoft
Virtualization type:                  full
L1d cache:                            432 KiB (9 instances)
L1i cache:                            576 KiB (9 instances)
L2 cache:                             18 MiB (9 instances)
L3 cache:                             18 MiB (1 instance)
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] intel_extension_for_pytorch==2.6.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.2.4
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] onnxruntime==1.21.0
[pip3] torch==2.6.0
[pip3] triton==3.2.0
[conda] Could not collect
```

### LINKED ISSUE #174049: [inductor] sdpa pattern matcher ignores scale and get wrong result
### 🐛 Describe the bug

In this issue #151098, the reproducer use a wrong scale. sdpa pattern matcher ignores the scale from user and use the default one (sqrt(1/64))
```python
#copied from original issue
import torch

class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.inv_scale = 1.0 / 8 ** 0.5
        self.query = torch.nn.Linear(64, 64)
        self.key = torch.nn.Linear(64, 64)
        self.value = torch.nn.Linear(64, 64)

    def forward(self, x1, attn_mask):
        q = self.query(x1).permute([0, 2, 1, 3])
        k = self.key(x1).permute([0, 2, 1, 3])
        v = self.value(x1).permute([0, 2, 1, 3])
        t1 = torch.matmul(q, k.transpose(-2, -1))
        t2 = t1.div(self.inv_scale) # wrong here
        t3 = t2 + attn_mask
        t4 = t3.softmax(dim=-1)
        t5 = t4.matmul(v)
        return t5
    
    def sdpa_fwd(self, x1, attn_mask):
        q = self.query(x1).permute([0, 2, 1, 3])
        k = self.key(x1).permute([0, 2, 1, 3])
        v = self.value(x1).permute([0, 2, 1, 3])
        return torch.nn.functional.scaled_dot_product_attention(
            q,
            k,
            v,
            attn_mask=attn_mask,
            dropout_p=0.0,
            is_causal=False,
            scale = 1 / self.inv_scale
        )


func = Model().to('cpu')


x1 = torch.randn(1, 16, 64, 64).to('cpu')

attn_mask = torch.zeros(1, 1, 16, 16).to('cpu')

test_inputs = [x1, attn_mask]

opt = torch.compile(func)

out_normal = func(x1, attn_mask)
out_sdpa = func.sdpa_fwd(x1, attn_mask)
out_opt = opt(x1, attn_mask)
print("wrong scale")
print("eager", out_normal[0,0,0,0])
print("eager sdpa", out_sdpa[0,0,0,0])
print("compile", out_opt[0,0,0,0])

func.inv_scale = 64 ** 0.5
opt = torch.compile(func)
print("default scale")
out_normal = func(x1, attn_mask)
out_sdpa = func.sdpa_fwd(x1, attn_mask)
out_opt = opt(x1, attn_mask)
print("eager", out_normal[0,0,0,0])
print("eager sdpa",out_sdpa[0,0,0,0])
print("compile", out_opt[0,0,0,0])
```

output:
```
wrong scale
eager tensor(0.4660, grad_fn=<SelectBackward0>)
eager sdpa tensor(0.4660, grad_fn=<SelectBackward0>)
compile tensor(0.0763, grad_fn=<SelectBackward0>)
default scale
eager tensor(0.0763, grad_fn=<SelectBackward0>)
eager sdpa tensor(0.0763, grad_fn=<SelectBackward0>)
compile tensor(0.0763, grad_fn=<SelectBackward0>)
```

We see result of torch.compile ignores the change of scale, although user code in eager is wrong

@eellison I think previous pr #172951 is just a workaround to close pattern matcher. We should fix pattern matcher for #151098

### Versions

main branch

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by jianyizh**
Fix #174049 
Fix #151098
scale in sdpa is passed as a scalar into pattern matcher. Some sdpa pattern replacement graph use the default value (sqrt(head_dim)), but eager code may not be default. In this case, there will be accuracy issue for torch compile. I add a check for scalar values. We may add support for non-default scale in future if we meet such cases in real workload (maybe follow the way in dropout)

We may revert #172951 before this pr because it tries to solve the same issue by disabling pattern matcher.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 1]
```diff
@@ -888,6 +891,39 @@ def fn(match):
...
+                    and len(node_args) > 1
+                    and isinstance(node.args[1], (float, int))
+                ):
+                    if inv_scale_factor is None:
+                        inv_scale_factor = node.args[1]
+                    else:
+                        assert inv_scale_factor == node.args[1]
+            query_node = match.kwargs["query"]
+            expected_inv_scale = math.sqrt(query_node.meta["val"].size(-1))
+            if expected_inv_scale != inv_scale_factor:
```
Comment (guangyey) [inline torch/_inductor/fx_passes/fuse_attention.py]: Eager mode could accept the scalar parameter. However, the current inductor does not respect the scalar parameter. So, add this check to explicitly match the attention pattern. The final solution should make the inductor respect the given scalar parameter?

Comment (jianyizh) [reply torch/_inductor/fx_passes/fuse_attention.py]: yes

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 2]
```diff
@@ -888,6 +891,39 @@ def fn(match):
...
                 return False
+        if check_scale_value:
+            has_scale = any("scale" in k for k in match.kwargs)
+            assert not has_scale, (
+                match.kwargs,
+                "do not check scale value when it's in kwargs",
+            )
+            div_nodes = filter_nodes(match.nodes, aten.div.Tensor)
+            inv_scale_factor: Optional[float] = None
+            for div_node in div_nodes:
```
Comment (eellison) [inline torch/_inductor/fx_passes/fuse_attention.py]: The scalar_workaround mechanism already exists for exactly this purpose (it's how dropout_p is handled in patterns 1-4). Instead of rejecting non-standard scales, we should capture the scale as a KeywordArg('inv_scale') via scalar_workaround and forward it to SDPA with scale=1.0 / inv_scale. This way non-standard scales fuse correctly rather than being refused.

**Review by eellison — COMMENTED** [review]
Thanks for the PR - can we use scalar_workaround for scale instead ?

File: torch/_inductor/fx_passes/fuse_attention.py (hunk 2 above)
Comment (jianyizh) [reply torch/_inductor/fx_passes/fuse_attention.py]: Thanks. I have updated to use scalar_workaround

**Review by eellison — APPROVED** [review]
